src/app/clusters/joint-fabric-datastore-server/joint-fabric-datastore-server.cpp

```diff
@@ -127,43 +182,142 @@ bool emberAfJointFabricDatastoreClusterAddGroupCallback(
     return true;
 }
 
-// TODO
 bool emberAfJointFabricDatastoreClusterAddKeySetCallback(
     CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
     const JointFabricDatastoreCluster::Commands::AddKeySet::DecodableType & commandData)
 {
+    CHIP_ERROR err                                                                              = CHIP_NO_ERROR;
+    JointFabricDatastoreCluster::Structs::DatastoreGroupKeySetStruct::DecodableType groupKeySet = commandData.groupKeySet;
+    JointFabricDatastorage & JointFabricDatastorage = Server::GetInstance().GetJointFabricDatastorage();
+
+    VerifyOrExit(JointFabricDatastorage.IsGroupKeySetEntryPresent(groupKeySet.groupKeySetID) == false,
+                 err = CHIP_ERROR_INVALID_ARGUMENT);
+    SuccessOrExit(err = JointFabricDatastorage.AddGroupKeySetEntry(groupKeySet));
+
+exit:
+    if (err == CHIP_NO_ERROR)
+    {
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Success);
+    }
+    else
+    {
+        ChipLogError(DataManagement, "JointFabricDatastoreCluster: failed with error: %" CHIP_ERROR_FORMAT, err.Format());
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::ClusterStatusCode(err));
+    }
+
     return true;
 }
 
-// TODO
 bool emberAfJointFabricDatastoreClusterRemoveNodeCallback(
     CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
     const JointFabricDatastoreCluster::Commands::RemoveNode::DecodableType & commandData)
 {
+    NodeId nodeId = commandData.nodeID;
+
+    if (commandPath.mEndpointId != kRootEndpointId)
```

Comment:
I wouldn't expect the JS Datastore to be on the root endpoint so I don't think this check is correct

Reply:
The check has been removed in the latest commit.